### ISSUE TITLE
Offline support

### ISSUE THREAD

**Issue by ErikMichelson** [hedgedoc/hedgedoc#2877]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
While being in offline mode, you still may create notes which will be stored by the service worker on your local device and which will be uploaded to the backend server the next time you have an internet connection.  
Optionally: Pinned notes will be downloaded and cached regularly by the service worker so that you can read (not edit!) them when the device is offline.

**Describe alternatives you've considered**
The alternative would be to have absolutely no offline support like the current version.

**Comment by tionis** [comment]
This feature request only handles cached read-only notes. Shouldn't writes, that are sent when possible (so eventual consistent) also be easy to implement as hedgedoc seems to be using y.js for crdts now?
My use case would be for tracking changes to a todo list in a unreliable network

**Comment by ErikMichelson** [comment]
That could be a next step yes, but if we try to integrate all these features at once, 2.0 will never be finished.
Having HedgeDoc to be resistant against online/offline -switching scenarios (like in a train) is actually something I'd like to see personally as well.

**Comment by tionis** [comment]
> That could be a next step yes, but if we try to integrate all these features at once, 2.0 will never be finished.

Yes of course, should I open a separate issue so it's tracked somewhere but can be pushed back to be looked at again after the 2.0 release?

**Comment by ErikMichelson** [comment]
I clarified this issue description to act as a meta-issue, so we can create sub-issues for implementing the single features.

**Comment by danthonywalker** [comment]
When using HedgeDoc on the go, often my Internet cuts in and out. It's very annoying when it reconnects that all the changes I've done are just erased; it borderline makes HedgeDoc unusable in a mobile context. +1 for this feature.

**Comment by jmtemmos** [comment]
+1 for this feature

**Comment by mikul9** [comment]
Without this feature, walking around a store and checking off things as you purchase them will not sync back to the server making it semi-useless for shopping lists.
